pkg/chunk/grpc/storage_client.go

```diff
@@ -0,0 +1,124 @@
+package grpc
+
+import (
+	"context"
+	"io"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+)
+
+type StorageClient struct {
+	cfg        Config
+	schemaCfg  chunk.SchemaConfig
+	client     GrpcStoreClient
+	connection *grpc.ClientConn
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (*StorageClient, error) {
+	grpcClient, conn, err := connectToGrpcServer(cfg.Address)
+	if err != nil {
+		return nil, err
+	}
+	client := &StorageClient{
+		cfg:        cfg,
+		schemaCfg:  schemaCfg,
+		client:     grpcClient,
+		connection: conn,
+	}
+	return client, nil
+}
+
+func (s *StorageClient) Stop() {
+	_, err := s.client.Stop(context.Background(), &Nothing{})
+	if err != nil {
+		return
+	}
+}
+
+// PutChunks implements chunk.ObjectClient.
+func (s *StorageClient) PutChunks(ctx context.Context, chunks []chunk.Chunk) error {
+	chunksInfo := &Chunks{}
+	for i := range chunks {
+		buf, err := chunks[i].Encoded()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		key := chunks[i].ExternalKey()
+		tableName, err := s.schemaCfg.ChunkTableFor(chunks[i].From)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		writeChunk := &Chunk{
+			Buf:       buf,
+			Key:       key,
+			TableName: tableName,
+		}
+
+		chunksInfo.Chunks = append(chunksInfo.Chunks, writeChunk)
+	}
+
+	_, err := s.client.PutChunks(ctx, chunksInfo)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func (s *StorageClient) DeleteChunk(ctx context.Context, chunkID string) error {
+	chunkInfo := &ChunkID{ChunkId: chunkID}
+	_, err := s.client.DeleteChunks(ctx, chunkInfo)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (s *StorageClient) GetChunks(ctx context.Context, input []chunk.Chunk) ([]chunk.Chunk, error) {
+	chunks := &Chunks{}
+	chunks.Chunks = []*Chunk{}
+	var err error
+	for _, inputInfo := range input {
+		chunkInfo := &Chunk{}
+		// send the table name from upstream gRPC client as gRPC server is unaware of schema
+		chunkInfo.TableName, err = s.schemaCfg.ChunkTableFor(inputInfo.From)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		chunkInfo.Key = inputInfo.ExternalKey()
+		chunks.Chunks = append(chunks.Chunks, chunkInfo)
+	}
+	streamer, err := s.client.GetChunks(ctx, chunks)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var result []chunk.Chunk
+	for {
+		receivedChunks, err := streamer.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		if receivedChunks != nil {
```

Comment:
You can remove this if and use `range receivedChunks.GetChunks()` in the next line.